### DISCUSSION TITLE
How to initialize package's development environment

### DISCUSSION THREAD

**Discussion by jtomori** [AcademySoftwareFoundation/rez#1424] [Q&A] [answered]
Hello,
rez newbie here.

First I'd like to thank you for developing this. I've been migrating some of my workflows to rez recently and it's been going pretty well.

I'm still figuring out what's the best way of using rez - developing, deploying, runtime environments.

One thing I am unsure about is how to use rez to create a package development environment - one where I can interactively develop, test, tweak package in package's git repo, e.g. in a development branch.

For more context, I'm mostly looking at develpoing Python packages and DCC-specific assets, like Houdini's HDAs. I use `rez build` to do things like build docs, do some cleanup, optimize some things, etc.

I _think_ I'm looking for something like `rez env package.py`, which would drop me into a env with all of the dependencies and env vars specified in the `package.py`.

Another workflow, although not ideal, would be to do a local build, then resolve that environment, do changes in it and then copy those changes back to the git repository. This involves some extra work, so I'd like to avoid it if possible.

The closest thing I've found is to do `rez build --scripts` and then source the `build-env.cmd`, which drops me into a "build" shell, with all dependencies there. This however doesn't set env vars in the `package.py`, which are needed, if I'm to work on Houdini HDAs.

A manual workaround for me would be to create a "env script" I'd source before the development. This script would contain the same env vars as rez's `package.py`, but I'd like to avoid maintaining identical env vars in two files.

I've also tried running `rez env --build`, but this dropped me into an empty rez context, not sure about the intended usage.

Hope my question makes sense to you. I secretly hope there's some simple solution for this, or a flag I've missed.

I use rez 2.111.3 on Windows.

Best,
Juraj

**Comment by jtomori** [comment]
Still wrapping my head around this.

After a bit of reading it looks like I'd like to resolve into an unbuilt package (where I'd do the development, in package's git repo). It'd be something like pip's [editable install](https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs).

Another avenue I'm exploring to do so is prepending my project's working directory into `local_packages_path`. This is currently failing because I don't have the expected folder structure: `pkg/version/package.py`. I only have `pkg/package.py`. Maybe some symlinking setup would help, will check how that works on Windows.

Am I approaching this correctly? It looks like many people do a local build with `rez build --clean --install` first. This makes sense with compiled code, but not during development with Python in my opinion, or Houdini (where I need to work on assets in an already resolved environment).

Some notes for my future self:

- https://github.com/AcademySoftwareFoundation/rez/issues/1263, a related topic
- https://youtu.be/rPe6JpzE_Nc, demo about isolated rez envs (@ColinKennedy, any chance to share more videos? I'd be really helpful to see others workflows.)

**Comment by jtomori** [comment]
Would anybody be willing to share their Python package development workflow?

Let's say that I have a package that depends on numpy. I specify that in the `requires` section of my `package.py`. To do the development, run tests, etc. I need to be in an environment with numpy available. So I could do `rez env numpy` (after having installed numpy with `rez pip`).

From that environment I can do what I need - editing files in package's repository, running tests etc. If I add more dependencies, then I need to update both `requires` and my `rez env` command, so this is not ideal if the list grows. Furthermore if I have set environment variables in the `commands` section of my package then I need to somehow recreate it again for my dev env. I hope you can see where I'm getting here.

**Comment by flord** [comment, marked as answer]
On linux, the way I used to do it is to simply symlink the local package content back to the source code directly. This can be done in the build script of the package and triggered only on local builds. This way, you can `rez-env yourTool` and stay inside this environment to run the tool you are developping as often as you want, and this is updated everytime you hit the save button in your IDE. 
On Windows however, I have yet to figure out how to achieve a similar workflow.

**Reply by jtomori** [reply]
Ooh, this is exactly what I was looking for.

I've tried it manually and it _seems_ to work well (even on Windows). Will update my build script and test it some more.

Thanks for this, appreciate your help.

**Reply by jtomori** [reply]
Thanks again, this works seamlessly.

During build I check whether it is a local installation (compare `REZ_BUILD_INSTALL_PATH` with `REZ_LOCAL_PACKAGES_PATH` in my setup), after which I symlink repository contents instead of copying them. During release I copy files as usual.
